README.ideal.md

```diff
@@ -0,0 +1,131 @@
+
+## Getting started
+
+To start using `nickel-nix`, you need `Nix` to be installed and configured.
+If it isn't already the case, you can get it with:
+
+```console
+$ curl -L https://nixos.org/nix/install | bash
+# We also need a couple of experimental Nix features
+$ mkdir -p ~/.config/nix
+$ echo 'experimental-features = nix-command flakes' >> ~/.config/nix/nix.conf
+```
+
+Then bootstrap a project with:
+
+```console
+$ nix flake init -t github:nickel-lang/nickel-nix
+# Edit the project file to fit your needs
+$ $EDITOR project.ncl
+# Enter the environment
+$ . .nickel-nix/activate
+```
+
+## Managing your environment with Nixel
+
+### Dependencies
+
+Nixel can be used to declare the dependencies for your project.
+These can then be instantiated using [Nix](https://nixos.org/nix).
+
+> TODO: Document
+
+### CI configuration
+
+You can use Nixel to generate a CI configuration.
+This is extremely handy if you want to test some non-trivial matrix of platforms/configuration.
+
+Only Github actions is supported at the moment, but we plan to add more options soon.
+
+> TODO: Document
+
+### Development services
+
+Nixel can also be used for some simple service management.
+In combination with the Nix integration, this allows you to quickly spawn any service that you might depend on.
+
+Once the services have been defined in your `project.ncl` file, you can run `nixel up` to start them.
+
+> TODO: Document
+
+### Deployment
+
+If you want to push things yet a bit further, Nixel can also interoperate with [tf-ncl](https://github.com/tweag/tf-ncl) to allow you to manage the deployment infrastructure for your project.
+
+> TODO: Document
+
+## Structure of the `project.ncl` file
+
+`project.ncl` is the main entrypoint for `nickel-nix`.
+It describes the packages that should be included in your development environment, as well as the template for any file you'd like to generate:
+
+```nickel
+let Nixel = import ".nickel-nix/lock.ncl" in
+{
+  nix_inputs.spec = {
+    pandoc = {},
+    make = {},
+    prettier = {},
+    redis = {},
+  },
+  shells.build =
+    {
+      packages = [nix_inputs.pandoc],
+      scripts = {
+        build = nix-s%"
+                #!/usr/bin/env bash
+
+                %{nix_inputs.make}/bin/make -j$(nproc) -l$(nproc)
+            "%,
+      },
+    }
+    & Nixel.shells.Rust
+    & Nixel.shells.Nickel,
+  shells.dev =
+    shells.build
+    & {
+      env = {
+        DEV_ENDPOINT = "http://localhost:1234",
+      },
+      packages = [nix_inputs.prettier],
+    },
+  services = {
+    postgresql = Nixel.services.postgresql,
+    redis.start = nix-s%"
+            %{nix_inputs.redis}/bin/redis
+        "%,
+  },
+  ci =
+    let CI = Nixel.CI.Gha in
+    {
+      type = "gha",
+      jobs.build = {
+        systems = [
+          { os : "ubuntu-latest" },
+          { os : "macos-latest" }
+        ],
+        matrix = {
+          system | { os : Str },
+          steps = [
+            CI.steps.checkout,
```

Comment:
That's indeed what @vkleen did for tf-ncl, e.g. factorizing the steps common to several workflows (https://github.com/tweag/tf-ncl/blob/main/ncl/github/setup-steps.ncl)

Reply:
Yet, the default is to have the steps. You can use a function, or have field that you can set to `false` to deactivate the generation of the default steps.

Reply:
I've added a `CI.standardSteps` variable for now. It's a bit sad because it would be great to use merging for this, but it doesn't play well with lists (and I think we really want a list here)

Reply:
I think I'll still object. This is gratuitous boilerplate. It's a line of code that _you must not forget to write_ in every shell file (with like 1% excepted). We should not consider acceptable that the default is more characters than the less common case.

Reply:
You might be right. I'm quite wary of prematurely trying to abstract over something as common as GitHub actions, as it has the huge advantage that most people will already be used to it.

Maybe we could have two layers: One (mostly autogenerated by https://github.com/nickel-lang/json-schema-to-nickel) that would map 1-1 to Gha, and one more opinionated building on top of it that tries to just work™. But in that case the high-level schema might be even more different from the original